### PR TITLE
stm32/sdram: Enforce gcc optimisation in sdram_test.

### PR DESCRIPTION
**Pull request by andrewleech**
Ensures consistent behavior and resolves the D-Cache bug at O0

Implementation adapted directly from investigation by @iabdalkader and @dpgeorge in https://github.com/micropython/micropython/pull/7841#issuecomment-925467003

If I understant correctly, the d-cache clean (flush) that's included in the d-cache disable should work correctly whenever the opt level is _not_ zero, so enforcing it to anything higher should work.

With exhaustive manually turned on in main.c I do get the ~6 second startup delay which seems to mirror the testing by @iabdalkader where it's faster than opt=0 but certainly is running a lot more code than with exhaustive=false

I enabled `MICROPY_HW_SDRAM_TEST_FAIL_ON_ERROR` and then "broke" one of my sdram settings:
`MICROPY_HW_SDRAM_COLUMN_BITS_NUM  9` -> `MICROPY_HW_SDRAM_COLUMN_BITS_NUM  10` => __fatal_error("Address bus overlap...")
So that section at least definitely works. 

I tried knocking out some other settings, eg 
``` c
    mp_hal_pin_config_alt_static_speed(MICROPY_HW_FMC_A4, MP_HAL_PIN_MODE_ALT, MP_HAL_PIN_PULL_NONE, MP_HAL_PIN_SPEED_VERY_HIGH, STATIC_AF_FMC_A4);
    // mp_hal_pin_config_alt_static_speed(MICROPY_HW_FMC_A5, MP_HAL_PIN_MODE_ALT, MP_HAL_PIN_PULL_NONE, MP_HAL_PIN_SPEED_VERY_HIGH, STATIC_AF_FMC_A5);
    mp_hal_pin_config_alt_static_speed(MICROPY_HW_FMC_A6, MP_HAL_PIN_MODE_ALT, MP_HAL_PIN_PULL_NONE, MP_HAL_PIN_SPEED_VERY_HIGH, STATIC_AF_FMC_A6);

```
and naturally that just triggers the same Address bus overlap error from sdram_test().

strangely though
``` c 
    mp_hal_pin_config_alt_static_speed(MICROPY_HW_FMC_D1, MP_HAL_PIN_MODE_ALT, MP_HAL_PIN_PULL_NONE, MP_HAL_PIN_SPEED_VERY_HIGH, STATIC_AF_FMC_D1);
    // mp_hal_pin_config_alt_static_speed(MICROPY_HW_FMC_D2, MP_HAL_PIN_MODE_ALT, MP_HAL_PIN_PULL_NONE, MP_HAL_PIN_SPEED_VERY_HIGH, STATIC_AF_FMC_D2);
    mp_hal_pin_config_alt_static_speed(MICROPY_HW_FMC_D3, MP_HAL_PIN_MODE_ALT, MP_HAL_PIN_PULL_NONE, MP_HAL_PIN_SPEED_VERY_HIGH, STATIC_AF_FMC_D3);
```
Didn't get picked up by exhaustive sdram_test, it does fail later though in `init_flash_fs(state.reset_mode)`  with a `MemManage_Handler()` -> `__fatal_error("MemManage");`

### CODE REVIEW

**Comment by iabdalkader** [comment]
Hi, this is a good way to test it, thanks! Note it doesn't fail with D2 disabled and `O0` either, I think it's because the pattern is 0xaa (10101010) so D2 is 0 anyway if it's not configured. If you try with any other bit that's 1 (like D1 or D3) it fails, so this fix is good.

File: ports/stm32/sdram.c [hunk 1]
```diff
@@ -310,8 +310,8 @@ bool __attribute__((optimize("O0"))) sdram_test(bool exhaustive) {
 
     // Test data bus
     for (uint32_t i = 0; i < MICROPY_HW_SDRAM_MEM_BUS_WIDTH; i++) {
-        *((uint32_t *)mem_base) = (1 << i);
-        if (*((uint32_t *)mem_base) != (1 << i)) {
+        *((volatile uint32_t *)mem_base) = (1 << i);
+        if (*((volatile uint32_t *)mem_base) != (1 << i)) {
             #if MICROPY_HW_SDRAM_TEST_FAIL_ON_ERROR
             snprintf(error_buffer, sizeof(error_buffer),
                 "Data bus test failed at 0x%p expected 0x%x found 0x%lx",
```
Comment (iabdalkader) [inline ports/stm32/sdram.c:317]: @andrewleech Should probably add `volatile` in this printf too for mem_base.

**Comment by andrewleech** [comment]
Ah thanks @iabdalkader yeah I thought the data checks should have picked it up already because it alternates between aa and 55
That `__DSB` is a great idea, I'd forgotten about that. I've pushed your diff here with the commit labelled with your identifier, hope that works for you?

**Comment by andrewleech** [comment]
I can also confirm this latest version fixes the databus error test when I comment out random data bus config lines in sdram.c

**Comment by iabdalkader** [comment]
> I've pushed your diff here with the commit labelled with your identifier, hope that works for you?

Yes sure, thanks!

**Pull request closed without merging**